### PR TITLE
Control, set_global_position behaves like get_global_position

### PR DESCRIPTION
**Pull request by toger5**
**get_global_position** behaved different than the according set function. (regarding rotations)

this lead to the annoying problem that:
```
rect_global_position = rect_global_position 
#of course the example is dumb, but there are reasons to do 
rect_global_position = rect_global_position * 0.5
```
moved the node (when it was rotated with a rotation pivot != (0,0) ).

screen shot 2018-07-11 at 10 40 33

In the case above:
### before:
 - get_global_position returned the green point
 - set_global_position used the red one
### now:

now both functions use the red one.
so setting with the same value does not move the node anymore.

## Second commit
also fixed behaviour when editing the pivot. Before it was changing the position of the node.
(the pivot stuff was really complex anyways)

I would like to have a review by @groud because I think you know fairly well how the control stuff should behave.

### CODE REVIEW

File: scene/gui/control.cpp [hunk 1]
```diff
@@ -128,10 +128,7 @@ bool Control::_edit_use_rotation() const {
 }
 
 void Control::_edit_set_pivot(const Point2 &p_pivot) {
-	Vector2 delta_pivot = p_pivot - get_pivot_offset();
-	Vector2 move = Vector2((cos(data.rotation) - 1.0) * delta_pivot.x - sin(data.rotation) * delta_pivot.y, sin(data.rotation) * delta_pivot.x + (cos(data.rotation) - 1.0) * delta_pivot.y);
-	set_position(get_position() + move);
-	set_pivot_offset(p_pivot);
+	set_pivot_offset(get_transform().xform(p_pivot));
```
Comment (groud) [inline scene/gui/control.cpp]: No this should definitely stay like it was. It is normal that in the _edit_set_pivot function that the node's position is changed. With your solution, if you set the pivot via the editor (with the V key), the node moves on the screen.
It's something that is less UX friendly than before.

File: scene/gui/control.cpp [hunk 2]
```diff
@@ -1705,7 +1702,13 @@ Size2 Control::get_end() const {
 
 Point2 Control::get_global_position() const {
 
-	return get_global_transform().get_origin();
+	Transform2D inv;
+
+	if (data.parent_canvas_item) {
```
Comment (groud) [inline scene/gui/control.cpp]: I think this is not a good solution to solve the problem. Getting the parent's transform is not a good solution to retrieve the global transform (mostly because of canvas layers). You should stick to use the node's get_global_transform() and get_transform() for that (not the parent's one).  

I'll try using a simpler solution.

**Review by groud — CHANGES_REQUESTED** [review]

File: scene/gui/control.cpp (hunk 2 above)
Comment (groud) [reply scene/gui/control.cpp]: My bad, it's done that way in the set_global_position(). So I guess it's ok.

Comment (groud) [reply scene/gui/control.cpp]: Anyway, I would simply replace all of this by a simple :
```
return (get_global_transform() * get_transform().affine_inverse()).xform(get_position())
```
Which is far easier to read.

File: scene/gui/control.cpp (hunk 1 above)
Comment (toger5) [reply scene/gui/control.cpp]: I was thinking about this a lot.
To some extend I agree. It is more obvious in the first place. changing pivot leaves the node at the same position and the rotation pivot is different after the operation.

But I feel like changing the pivot really only should change the pivot, to reflect what changing pivot does when doing it manually in code.
I guess most ppl need a little bit of time to wrap their heads around the pivot/position concept and why the objects seems to move when the pivot is changed and the rotation != 0.

So it is helpful when the editor already represents that behaviour. Playing around with it makes it really easy to figure out how the pivot behaves and you learn to deal with it. when changing the pivot in game you already know what you do and are not completely confused.

It also can lead to
TLDR;
the editor should represent api pivot operations.

File: scene/gui/control.cpp (hunk 2 above)
Comment (toger5) [reply scene/gui/control.cpp]: It's definitely nicer. But you need to be more comfortable with matrix multiplication.
I can change it though. shouldn't be touched by end users anyways.

File: scene/gui/control.cpp (hunk 1 above)
Comment (groud) [reply scene/gui/control.cpp]: No, I don't think it should.
The canvas item editor is a visual tool, it is meant to be easy to use, with expected visible results. If the node moves when you move the pivot it definitely complexifies the workflow, and makes things hard to understand.

To sum up my point: if you use visual tools, the result should be as obvious as possible, even if it screws up the position or other values. This should obviously not happen when you use the inspector, where changing the pivot should not change the position.

File: scene/gui/control.cpp (hunk 2 above)
Comment (groud) [reply scene/gui/control.cpp]: Well, you need to be comfortable with transforms if you want to touch this part of the code I guess. So I think is this situation, anything that reduce the number of lines of code is good.

File: scene/gui/control.cpp (hunk 1 above)
Comment (toger5) [reply scene/gui/control.cpp]: Okay.
So the '_edit' function. Is doing it's special thing with fixing the position after pivot change.
But changing only pivot, only changes pivot!
Thats sounds reasonable!!

Comment (groud) [reply scene/gui/control.cpp]: Exactly :)

**Comment by akien-mga** [comment]
What's the status on this PR?

**Comment by groud** [comment]
> What's the status on this PR?

The requested changes still need to be made. Also, note that this is breaking compatibility change.

**Comment by toger5** [comment]
This should be done before 3.1 right?

Might look into it today...

**Comment by aaronfranke** [comment]
@toger5 Just a reminder about this. The existing behavior as you showed is clearly wrong, so something should indeed be done about it, and we would appreciate the fix.

**Comment by toger5** [comment]
I Don't totally get what the ideal solution @groud has in mind.
I agree this is important to get merged.

I currently also Don't have the time to work on it.
Maybe groud has time for a fork of the pr including his proposed change of the node position when changing the pivot?

**Comment by akien-mga** [comment]
To be fair, I don't really understand what this PR aims to solve exactly, and since it's breaking compatibility, it couldn't be merged as is for 3.2.

I would suggest opening a proposal with a detailed use case on https://github.com/godotengine/godot-proposals so that it can be discussed and approved before implementing - a priori for 4.0 if it must break compat.

**Comment by toger5** [comment]
@akien-mga yea makes sense to close it for now. Let me try to explain the issue if it didnt made sense to you. Basically the functions: `get_global_position` `set_global_position` behave different and return other cooridnates then the control is rotated.
The result is, that when calling `set_global_position(pos1)` `get_global_position()` does not necassarly return `pos1` because one is correcting for rotation and the other function is not.
